base-files: sysupgrade: bkp/rstr/autoupg with pkgs

stumper932 opened the pull request:
This PR adds extra functionality to **sysupgrade** by **automating the upgrade process**. As a bonus it will also backup and restore all packages.

Whats new:
2018-10-15 https://peertube.video/videos/watch/8417d691-80f5-4c73-b3c9-652d61918fa3

1. Upgrade-command: **-U [\<url>]** 

- sets the upgrade \<url> (specific per device and needs to be set by the user once)
- without \<url> it checks if the set url is correct by spidering the OpenWRT webpage.

2. Upgrade-command: **-w**

- checks the OpenWRT webpage for the latest version and displays result.

3. Upgrade-command: **-Zzz** will perform an **autoupgrade of OpenWRT**. 

- checks for new version
- if an upgrade is available, downloads the upgrade file based on the url set with -U by the user and verifies with the checksum from the webpage. 
- starts itself with the upgrade file
- creates a package cache (matching the upgrade_id), and includes all installed packages.
- uninstalls all packages then pushes the cache to the overlay
- inserts a runonce service 'sysupgrade' into init.d and a link into rc.d 'S96sysupgrade'
- then resumes the standard upgrade process and reboots
- on boot starts S96sysupgrade
- S96sysupgrade will reinstall all packages from the packagelist
- if there is a local cache it will install from there otherwise it will try to reach the configured feed and use that instead.
- S96sysupgrade will remove itself after successful installation, or will disable itself after failure.
(see system log for success/failure)

4. Backup-command: **-b** 

- **saves** a package cache ** by **default**.
- if it can reach the package feed it will download and store a cache of all installed packages and opkg-lists into the backup.

5. Backup-command: **-r** 

- **restores** from package cache ** by **default**. 
- if there is only a packagelist (no local cache) and it can reach the package feed it will download and reinstall all packages from there.

6. Backup-command: **-k** 

- **lists** all installed user apps. 

_Todo:_
- sysupgrade directly from \<url>, this could be real nice for devices with less/low storage space.
- automatic scheduling with cron when setting the upgrade url. (any ideas on when? e.g. 04.30 every 3 days or something like that).

stumper932 commented:
Please review and test, thank you

bobafetthotmail commented:
On 8/20/2018 5:34 PM, Jeff Kletsky wrote:

For this to be part of the "default" sysupgrade process, I'd want it to handle situations where:

  *   Package names change
  *   Package dependencies change (including not installing "stale" dependencies)
  *   Packages are obsoleted or are not longer available

This is especially important across major versions, but can occur at any time.

I also don't see how permanent failures, as well as potentially transient failures are managed or logged and brought to the attention of the end user.

Further, this seems to assumes connectivity to the package server is available on first boot after upgrade, which is not always the case. This may be due, for example, to flashing a device using a non-production network, or certain packages required for rest-of-world connectivity.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/openwrt/openwrt/pull/1310#issuecomment-414360808>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AJDeo8FOaw0BTuTXwJ3zPzCgJgcEajEbks5uStb0gaJpZM4WEEvB>.

Imho if package names change or a package is obsoleted it should just not install that particular package.


Changes in package dependencies could be worked around a bit by not installing libraries (and relying on whatever the packages state they need at the moment of re-installation.

I also think it should at least wait until it can connect to the internet.

If the device lacks crucial packages for internet connectivity it's ok to wait too, as in any case the user would have to act and manually install things, but woul only need to install packages required for internet access.

Could be cool to specify some kind of local package server or something.


-Alberto

stumper932 commented:
@jeffsf

1. Package names change
    - Thats a problem
2. Package dependencies change (including not installing "stale" dependencies)
    - It only installs "direct user request" packages.
3. Packages are obsoleted or are not longer available
    - obsolete packages, i dont think they are around after the sysupgrade, and like i answered in 2.
    - not longer available would be the same as 1.
4. I also don't see how permanent failures, as well as potentially transient failures are managed or logged and brought to the attention of the end user.
    - opkg logs can be found in system log, under S95done.
5. Further, this seems to assumes connectivity to the package server is available on first boot after upgrade, which is not always the case. This may be due, for example, to flashing a device using a non-production network, or certain packages required for rest-of-world connectivity.
    - The install is a pending until okg update succeeds in rc.local, so you can reboot or manually execute rc.local. After success it removes itself from the system.

6. One "helpful" change in opkg would be identifying if a package is installed explicitly, or as a dependency of another. At least as I read the "help" for opkg, that information does not appear to be captured.
    - Its not a problem like i answered in 2, unless im missing something.


@bobafetthotmail 

1. Imho if package names change or a package is obsoleted it should just not install that particular package.
    - It only installs "direct user request" packages. I think it should echo the failed packages to e.g. /etc/config/.restore_failed or something.

2. Changes in package dependencies could be worked around a bit by not installing libraries (and relying on whatever the packages state they need at the moment of re-installation.
    - Like 1, dependencies are resolved by opkg, just like when you install manually.

3. I also think it should at least wait until it can connect to the internet.
    - It stops when opkg update fails, it will retry on next boot. You can also manually run rc.local.

4. If the device lacks crucial packages for internet connectivity it's ok to wait too, as in any case the user would have to act and manually install things, but woul only need to install packages required for internet access.
    - Like 3, manually install then run rc.local or reboot.

5. Could be cool to specify some kind of local package server or something.
    - That would make things, as i see it now: complicated. Space restriction on smaller devices could be a problem.

FangsMcWolf commented:
I like the idea of being able to have previous user-installed packages being made easily available again.  Perhaps in the Software page, a third tab that would list previously installed packages (that are not currently installed), along with any known information.  Is the package still known/available?  Does it have new (or no longer needed) dependencies?  Things like that.  If the name has been changed, it wouldn't be listed as known/available anymore.  If it's not in any current repositories, then again, not known/available.  If it is still using the same name and is current, then the user could optionally reinstall it, if they wish to continue using it.

When a sysupgrade is being processed, a list of installed software packages could be exported to a .conf file, and after the upgrade is done, that third tab could show what's missing.  In addition, could probably be used to show what is installed that wasn't previously installed.  The tab would only list differences from before the most recent upgrade, and the current version, not an accumulated list of changes.  Though an export option would be a good idea, in case someone wants a reminder of what they've used before a few versions ago.

stumper932 commented:
So i took all your advice to heart and added them into this update.
I cannot strike out the first comment, so let me know if i should alter it there.

Whatsnew: (https://streamable.com/qdvc2)

backup/sysupgrade/restore with packages offline and online
firmware upgrade can be scheduled with a cronjob after setup (-U)
it will detect when there is a new update (-Zzz) and perform
firmware upgrade.

added upgrade-option:
-U [<url>] test 'url', or set 'auto upgrade url' to *
-w what is the latest stable version *
-Zzz perform auto-upgrade, make sure the url is set *
(*) requires internet access

added backup command:
-k | --list-packages --- list currently installed user request packages.

Let me know what you think.

Leo-PL commented:
But still, this won't work for case in which you need extra packages just to establish WAN connectivity. In such case, sysupgrade "-f" option can be abused to preload required packages on overlay, so router boots to already configured state after migration, and has those extra packages available right after bootup.

stumper932 commented:
> But still, this won't work for case in which you need extra packages just to establish WAN connectivity. In such case, sysupgrade "-f" option can be abused to preload required packages on overlay, so router boots to already configured state after migration, and has those extra packages available right after bootup.

Actually it does, a local repo/cache is created. Packages will be installed from there on first boot.

stumper932 commented:
New update pushed, also edited the comments to reflect current state.

Leo-PL commented:
But will it support kernel modules also?
It is needed to preload the modules matching the new kernel after upgrade for them to load successfully after reboot.

stumper932 commented:
> But will it support kernel modules also?
- yes
> It is needed to preload the modules matching the new kernel after upgrade for them to load successfully after reboot.
- The **distfeed** is configured with the **new version_id**, therefore gets the **new package lists** which contains the information on packages and their **new dependencies**.

Just test your configuration out in a VM.

Leo-PL commented:
Sounds great! I'm actually willing to test this on one of my WDR4300's I use for development with Wireguard over the weekend :)

stumper932 commented:
New update pushed, fixed some bugs/typos.

stumper932 commented:
> Sounds great! I'm actually willing to test this on one of my WDR4300's I use for development with Wireguard over the weekend :)

How did it go?

Leo-PL commented:
Actually I waited a bit for stabilization, because you were pushing fixes all the time :)
I think it would be great if you can notify me that your patch is more-or-less ready for testing - then I'll include it in my next build.

stumper932 commented:
I added a testing branch, its on 17.01.5 because i configured it to only auto-upgrade to a stable version. Much got in the way today, especially logging. Ill send you a message when its done. Probably tomorrow =)

stumper932 commented:
> Actually I waited a bit for stabilization, because you were pushing fixes all the time :)
> I think it would be great if you can notify me that your patch is more-or-less ready for testing - then I'll include it in my next build.

Its done, debug logging is enabled so don't freak out ;) .. it will dump a lot of information to stdout.
logs can be found in: /tmp/log , there will be a pre and post log after upgrade succeeds.

1. The easiest approach unless you want to compile is to install 17.01.5 and replace /sbin/sysupgrade.
(if you want to compile, use the testing branch)
2. Install any packages you want to test with
3. Run sysupgrade -U <your device upgrade url from openwrt website>, it will filter out the release number automatically.
4. Run sysupgrade -Zzz and it will upgrade to the latest version (currently 18.06.1).
5. View logs in /tmp/log

I've tested this multiple times in a VM and the results are great.
Good luck and thanks for testing.

stumper932 commented:
Forked "luci" repository , fixed the system.lua to use sysupgrade instead of tar. (17.01)
You can use luci backup and restore with packages.

stumper932 commented:
Added GPG verification
Added support for snapshots (used to be releases only) (inlcudes kmods)
Updated PR text.

*No error messages, yet*
exit/return code = 0 everything is good ( # echo $? )
everything else, there is a problem

*Known bugs*
Currently there is a problem with opkg not fetching the correct ip candidate.
So you can end up with ip-tiny instead of ip-full or the other way around.
This is a problem with opkg afaik.

It seems i am done here, there is not much to do until people start testing.

I created three testing branches:
- testing-master
- testing-v18.06.1
- testing-v17.01.5

stumper932 committed:
backup/sysupgrade/restore with packages offline and online
firmware upgrade can be scheduled with a cronjob after setup (-U)
it will detect when there is a new update (-Zzz) and perform
firmware upgrade.

added upgrade-option:
-U [<url>] test 'url', or set 'auto upgrade url' to *
-w what is the latest stable version *
-Zzz perform auto-upgrade, make sure the url is set *
(*) requires internet access

added backup command:
-k | --list-packages --- list currently installed user request packages.

Signed-off-by: Walter van West <w.vanwest@ziggo.nl>

stumper932 commented:
Confirmed working on a TP-Link Archer C7 v2

Jookia commented:
Works on the WNDR3800, but doesn't auto-install luci when installing a snapshot.
It also didn't start my 6in4 interface after installing the package.

[sysupgrade.8.log](https://github.com/openwrt/openwrt/files/2603608/sysupgrade.8.log)
[sysupgrade.9.log](https://github.com/openwrt/openwrt/files/2603610/sysupgrade.9.log)

stumper932 commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -135,6 +165,632 @@ add_overlayfiles() {
+
+#test_url(){	wget -q -s "$1";return $?;}
+test_url(){
+	local result=`wget -q -s "$1" >/dev/null 2>&1;echo $?`
```
My bad, i did not know wget was a link to ufetch-client. When the wget packge from the repo is installed it replace the link. Options do not translate 1:1. So i changed wget into ufetch-client.
This should resolve the issue. I updated the working-master branch. (as you probably noticed I have some other priorities atm).

Thanks for testing, if you find any other problems or have additional comments. Feel free to mail me personally or drop me a line.

ynezz commented:
> This PR adds extra functionality to sysupgrade

Please keep in mind, that sysupgrade in OpenWrt core should be slim as it should support a very broad range of targets, starting with support for devices having just 4M of total FLASH memory, leaving much less for a complete OS. So adding gpg as a hard dependency is a not an option.

> As a bonus it will also backup and restore all packages.

Ability to use binary packages is a nice OpenWrt feature, but it's actually just a by-product and it was never designed for automatic (unattended) system upgrades in mind, this is actually very hard topic, with a lot of corner cases etc.

> sysupgrade directly from <url>, this could be real nice for devices with less/low storage space.

There is already such option, I use it quite often:
 
Usage: /sbin/sysupgrade [<upgrade-option>...] <image file or [URL](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/sbin/sysupgrade#L296)>

> Upgrade-command: -Zzz will perform an autoupgrade of OpenWRT.

Take a look at https://github.com/aparcar/attendedsysupgrade-server there's also something similar in [Gluon](https://github.com/freifunk-gluon/gluon/tree/master/package/gluon-autoupdater).

So to sum it up, it's very unlikely, that is ever going to be merged in OpenWrt core, but if you would like to continue working on this, I think, that you can do that in a separate feed/package which could be then installed on top of the core OpenWrt.

ibex-are-goats commented:
I think it is a very big mistake for openwrt developers like @ynezz to ignore the regular openwrt user base who want progress. Stagnant openwrt development is one reason why LEDE happened in the first place and we don't want another split.

yurtesen commented:
@stumper932 This seems quite complicated for such a simple task. I may be way off, but can't this be simplified? I could download image myself for example. It seems like over-engineered?

Isn't it enough that `opkg` adds the name of the package to a list when user explicitly installs a package. It could be saved in /etc/opkg and can be kept between updates. Instead of sysupgrade trying to build this list and dependencies from installed packages at update? Why it can't be a static list of user installed packages?

After an update user can re-install missing packages at any point when the router has network using a command when he logs in. It does not have to be automatically executed (while this could be given to user as an option perhaps during upgrade).  System can compare existing packages and the ones in the pre-saved list and also check with `opkg` the available packages for the platform. Then compile a simple list like below on cli or a similar list in luci

pkg-a Not found for this release  [will be removed from list manual action needed]
pkg-b v1.2.3                               [reinstall]
pkg-c v.2.3.4                               [reinstall]
Continue (Y/n)?

Probably store the rejects in a different file so user can review them later.

When the packages are installed, the current dependencies would be resolved and installed. So that would be quite straightforward. If anything fails or missing it is user's responsibility to figure it out.

stumper932 commented:
I dont have any time atm, this is a one-time reply only. So fork it, experiment. Having that said:

> @stumper932 This seems quite complicated for such a simple task. I may be way off, but can't this be simplified? and be implemented with smaller modifications to `sysupgrade` ?

There are 3 functionalities:

1. Backup/Restore with packages.
2. Sysupgrade with packages. (differs from 1 because you need the newer packages)
3. GPG Verification.

There are other problems, not everyone has opkg. There are routers out there that have 4MB disk space. Everything needs to be done in /tmp and thus is gone after a reboot, this complicates matters.

> Isn't it enough that `opkg` adds the name of the package to a list when user explicitly installs a package. It could be saved in /etc/opkg and can be kept between updates. Instead of sysupgrade trying to build this list and dependencies from installed packages at update? Why it can't be a static list of user installed packages?

If thats what you want, just run  Backup-command: -k  lists all installed user apps.

> After an update user can re-install missing packages at any point when the router has network using a command when he logs in. It does not have to be automatically executed (while this could be given to user as an option perhaps during upgrade). System can compare existing packages and the ones in the pre-saved list and also check with `opkg` the available packages for the platform. Then compile a simple list like below on cli or a similar list in luci
> pkg-a Not found for this release [will be removed from list manual action needed]
> pkg-b v1.2.3 [reinstall]
> pkg-c v.2.3.4 [reinstall]
> Continue (Y/n)?
> 
> Probably store the rejects in a different file so user can review them later.

That's sort of what it does, but you cant trust on the fact that you have internet on the other side of an upgrade. You need to correctly download everything before you push the upgrade. And this fact complicates the process. To download all proper packages you need to have a clean package log.
The proper repo url, installed packages and their dependencies. Only then can you perform and upgrade.

> When the packages are installed, the current dependencies would be resolved and installed. So that would be quite straightforward. If anything fails or missing it is user's responsibility to figure it out.

Only if you are on the upgraded firmware, what if you dont have internet because of missing kmods ?
Not everyone has second router and it can be really annoying to manually insert those missing files.

Good luck. I hope all goes well.

yurtesen commented:
@stumper932 One has to accept there is no reasonable solution to all the problems. Trying to resolve all possible scenarios will result in over complicated scripts. 

I would think that if somebody does not have opkg and 4MB storage, he would either not have any extra packages installed or will be competent enough to reinstall themselves. (or build an image with required packages, compressed). You have to cut the line somewhere, not all devices can be supported. 

For example, I had a device with so little space that there was no way to install packages afterwards, I had to build image with the few extra packages which fit in compresed format. So your script would have crashed upgrade for sure.


About the internet connection, you are somewhat right, but also you have to understand that this person installed those packages once without internet. Can do it again. I have this problem and I simply build my own image with all packages I need for upgrades. But I understand not everybody wants to go through the trouble. 

Also my custom image causes your mods to not recognize the extra packages I have, over the default image from openwrt for my system (as they appear in "rom"). Therefore it would have failed if I tried to upgrade my system using default image and thinking your scripts would download all the necessary stuff. It would be pretty difficult to fix bug...

Yes, not everybody will have internet after upgrade. But it is a compromise for the first version of the solution to keep the solution simple. Over half of the time, people will have internet with base packages anyway. For downloading required packages you would need to check temp space and package sizes etc. to calculate if upgrade will succeed. 

The next step would be keeping store of installed packages on default images per device in repository (I am not sure if it exists now?) and also package sizes so opkg can calculate how much space is needed for pre-download packages to a temp location. Perhaps even a user defined temp location, or allow user to download them through web interface pre-installation and user can upload them back after upgrade.

Root-Core commented:
Sorry for necro posting.. but it's indeed sad that this still is an unresolved issue.
Is there any work in that direction? I hold back updates - for way too long - because of it.
